compendium/DeclarativeServices/test/TestDSConfigAdminOperationOrder.cpp

```diff
@@ -0,0 +1,162 @@
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=============================================================================*/
+
+#include "TestFixture.hpp"
+#include "gtest/gtest.h"
+
+#include "cppmicroservices/Constants.h"
+#include "cppmicroservices/ServiceObjects.h"
+
+#include "TestInterfaces/Interfaces.hpp"
+
+namespace test {
+/**
+   * Verify that DS must complete modification before deactivation of a component. 
+   */
+TEST_F(tServiceComponent, testModificationBeforeDeactivation) //DS_CAI_FTC_21
+{
+  // Start the test bundle containing the component name.
+  std::string componentName = "sample::ServiceComponentCA09";
+  cppmicroservices::Bundle testBundle = StartTestBundle("TestBundleDSCA09");
+
+  // Use DS runtime service to validate the component description and
+  // Verify that DS is finished creating the component data structures.
+  scr::dto::ComponentDescriptionDTO compDescDTO;
+  auto compConfigs =
+    GetComponentConfigs(testBundle, componentName, compDescDTO);
+  EXPECT_EQ(compConfigs.size(), 1ul) << "One default config expected";
+  EXPECT_EQ(compConfigs.at(0).state,
+            scr::dto::ComponentState::UNSATISFIED_REFERENCE)
+    << "component state should be UNSATISFIED_REFERENCE";
+
+  // Get a service reference to ConfigAdmin to create the component instance.
+  auto configAdminService =
+    GetInstance<cppmicroservices::service::cm::ConfigurationAdmin>();
+  ASSERT_TRUE(configAdminService) << "GetService failed for ConfigurationAdmin";
+
+  // Create configuration object
+  auto configObject = configAdminService->GetConfiguration(componentName);
+
+  // Update property
+  cppmicroservices::AnyMap props(
+    cppmicroservices::AnyMap::UNORDERED_MAP_CASEINSENSITIVE_KEYS);
+  const std::string instanceId{ "instance1" };
+  props["uniqueProp"] = instanceId;
+  configObject->Update(props);
+
+  // Start dependent bundle
+  auto depBundle = StartTestBundle("TestBundleDSTOI1");
+
+  // GetService to make component active
+  auto service = GetInstance<test::CAInterface1>();
+  ASSERT_TRUE(service) << "GetService failed for CAInterface";
+
+  compConfigs = GetComponentConfigs(testBundle, componentName, compDescDTO);
+  EXPECT_EQ(compConfigs.size(), 1ul) << "One default config expected";
+  ASSERT_EQ(compConfigs.at(0).state, scr::dto::ComponentState::ACTIVE)
+    << "Component state should be ACTIVE";
+
+  const std::string update1{ "update1" };
+  props["uniqueProp"] = update1;
+  configObject->Update(props);
```

Comment:
See Patty's changes from "update methods to return future"
auto fut = configuration->Update(props);
fut.get();

Reply:
I'm not doing anything with the returned future; in this test case, simply waiting and deactivating bundle in next line.
and validating that even though I'm not waiting on the future, DS does the modification first and Deactivation after that.